malariagen_data/ag3.py

```diff
@@ -7951,6 +7972,502 @@ def plot_roh(
+            sample_query=sample_query,
+            analysis=analysis,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        gt = allel.GenotypeDaskArray(ds_haps["call_genotype"].data)
+
+        ht = gt.to_haplotypes().compute()
+
+        calibration_runs = dict()
+        for window_size in tqdm(window_sizes, desc="Compute H12"):
+            h1, h12, h123, h2_h1 = allel.moving_garud_h(ht, size=window_size)
+            calibration_runs[str(window_size)] = h12
+
+        return calibration_runs
+
+    def plot_h12_calibration(
+        self,
+        contig,
+        analysis,
+        sample_query,
+        sample_sets,
+        cohort_size=20,
+        window_sizes=(100, 200, 500, 1000, 2000, 5000, 10000, 20000),
+        random_seed=42,
+        title=None,
+    ):
+        """Plot h12 GWSS calibration data for different window sizes.
+
+        Parameters
+        ----------
+        contig: str
+            Chromosome arm (e.g., "2L")
+        analysis : {"arab", "gamb_colu", "gamb_colu_arab"}
+            Which phasing analysis to use. If analysing only An. arabiensis, the
+            "arab" analysis is best. If analysing only An. gambiae and An.
+            coluzzii, the "gamb_colu" analysis is best. Otherwise, use the
+            "gamb_colu_arab" analysis.
+        sample_sets : str or list of str, optional
+            Can be a sample set identifier (e.g., "AG1000G-AO") or a list of
+            sample set identifiers (e.g., ["AG1000G-BF-A", "AG1000G-BF-B"]) or a
+            release identifier (e.g., "3.0") or a list of release identifiers.
+        sample_query : str, optional
+            A pandas query string which will be evaluated against the sample
+            metadata e.g., "taxon == 'coluzzii' and country == 'Burkina Faso'".
+        cohort_size : int, optional
+            If provided, randomly down-sample to the given cohort size.
+        window_sizes : int or list of int, optional
+            The sizes of windows used to calculate h12 over. Multiple window
+            sizes should be used to calibrate the optimal size for h12 analysis.
+        random_seed : int, optional
+            Random seed used for down-sampling.
+        title : str, optional
+            If provided, title string is used to label plot.
+
+        Returns
+        -------
+        fig : figure
+            A plot showing h12 calibration run percentiles for different window
+            sizes.
+
+        """
+
+        # get H12 values
+        calibration_runs = self.h12_calibration(
+            contig=contig,
+            analysis=analysis,
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            window_sizes=window_sizes,
+            cohort_size=cohort_size,
+            random_seed=random_seed,
+        )
+
+        # compute summaries
+        q50 = [
+            np.median(calibration_runs[window]) for window in calibration_runs.keys()
```

Comment:
Using `calibration_runs.keys()` here makes me nervous, because it relies on the keys always coming out in the same order as the `window_sizes` parameter was originally provided. This may in fact happen, but I would prefer to guarantee the correct order, e.g., by doing something like:

```suggestion
            np.median(calibration_runs[str(window)]) for window in window_sizes
```

...and similarly below.